File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/security/solver/AtomicLabelTerm.java
package edu.cornell.cs.apl.viaduct.security.solver;

import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice;
import edu.cornell.cs.apl.viaduct.algebra.solver.AtomicTerm;
import edu.cornell.cs.apl.viaduct.security.Label;
import edu.cornell.cs.apl.viaduct.security.Principal;

/** Atomic terms like constants and variables but joins. */
public abstract class AtomicLabelTerm extends LabelTerm {
  @Override
  public abstract AtomicLabelTerm confidentiality();

  @Override
  public abstract AtomicLabelTerm integrity();

  /** Term that corresponds to performing {@link Label#swap()}. */
  public abstract AtomicLabelTerm swap();

  /** Term that corresponds to the {@link Label#join(Label)} of {@code this} and {@code that}. */
  public abstract LabelTerm join(Label that);

  @Override
  abstract AtomicTerm<FreeDistributiveLattice<Principal>> getConfidentialityComponent();

  @Override
  abstract AtomicTerm<FreeDistributiveLattice<Principal>> getIntegrityComponent();
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/security/solver/AtomicLabelTerm.kt
package edu.cornell.cs.apl.viaduct.security.solver

import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice
import edu.cornell.cs.apl.viaduct.algebra.solver.AtomicTerm
import edu.cornell.cs.apl.viaduct.security.Label
import edu.cornell.cs.apl.viaduct.security.Principal

/** An atomic term such as a constant or a variable but a join. */
sealed class AtomicLabelTerm : LabelTerm() {
    abstract override fun confidentiality(): AtomicLabelTerm

    abstract override fun integrity(): AtomicLabelTerm

    /** The term that corresponds to performing [Label.swap]. */
    abstract fun swap(): AtomicLabelTerm

    /** The term that corresponds to the [Label.join] of `this` and [that]. */
    abstract fun join(that: Label): LabelTerm

    abstract override val confidentialityComponent: AtomicTerm<FreeDistributiveLattice<Principal>>

    abstract override val integrityComponent: AtomicTerm<FreeDistributiveLattice<Principal>>
}
